examples/scripts/blender/procedural_dataset_generator.py

```diff
@@ -0,0 +1,1326 @@
+from __future__ import annotations
+
+"""
+Blender exporter of SDF models with support for generating procedural datasets via node
+based modifiers, e.g. Geometry Nodes.
+"""
+
+# Please, manually adjust the last tested (and working) Blender version for this script
+LAST_TESTED_VERSION: Tuple(int, int, int) = (3, 1, 0)
+
+import enum
+import os
+import random
+import shutil
+import sys
+from os import path
+from types import ModuleType
+from typing import Any, Dict, Iterable, List, Optional, TextIO, Tuple, Union
+from xml.dom import minidom
+from xml.etree import ElementTree
+
+# Raise an informative error in case the script is run outside Python env of Blender
+try:
+    import bpy
+except ImportError as err:
+    raise ImportError(
+        "Python module of Blender 'bpy' not found! Please, execute this script inside "
+        "a Blender environment (e.g. copy-paste into Scripting tab)."
+    ) from err
+
+
+def print_bpy(msg: Any, file: Optional[TextIO] = sys.stdout, *args, **kwargs):
+    """
+    Helper print function that also provides output inside the Blender console,
+    in addition to the system console.
+    """
+
+    print(msg, file=file, *args, **kwargs)
+    for window in bpy.context.window_manager.windows:
+        for area in window.screen.areas:
+            if area.type == "CONSOLE":
+                bpy.ops.console.scrollback_append(
+                    {"window": window, "screen": window.screen, "area": area},
+                    text=str(msg),
+                    type="ERROR" if file == sys.stderr else "OUTPUT",
+                )
+
+
+class sdf_model_exporter(ModuleType):
```

Comment:
As far as I can tell, [sdf_exporter.py](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo6/examples/scripts/blender/sdf_exporter.py) exports the **entire scene as a static model** using a single `.dae` for all `<visual>` and `<collision>` entities. This script also supports exporting albedo textures, lights and light maps. In addition, selection of the export filepath can be done through GUI.

The export module of the script in this PR exports **selected objects (or all if none is selected) as a static/dynamic model**. If the model is dynamic, inertial properties can be automatically estimated. By default, `<visual>` is exported as a single `.dae`, whereas `<collision>` as a single `.stl` (can be extended to other formats). Before exporting mesh, a "pre-export hook" is called that allows the visual/collision geometry to be exported at different levels of detail (or otherwise modified). PBR textures are supported by specifying path to directory with PBR texture sets (if more than one texture set is present, it is sampled randomly). This script does not export lights (could be added - part of TODOs), nor albedo textures (I think baking textures might be more general and it would support procedural materials - part of TODOs). Light maps are not supported as I was targeting export of single models that can be combined later in Gazebo into worlds with dynamic lighting and interchanged at will.

I tried to keep the script as general as possible. It is designed in a way where it can be extended through inheritance (so that the procedural exporter does not need to re-implement everything). However, it might need to be completely refactored to support multiple use-cases. Implementing features from future work might make some things easier.

---

> I wonder if it'd be beneficial to refactor out a generic SDF exporter class that combines the two.

I think it is a good idea to have a generic SDF exporter (could be a Blender addon or just a Gazebo Python module).

Reply:
I'm not sure what to do about the SDF exporter yet - we should review the code before deciding that.

Regarding the export formats, I would suggest having a section in the tutorial describing the inputs and outputs of the script. An example directory structure of the output files would also be helpful, like the "Model Directory Structure" section in this tutorial https://gazebosim.org/api/gazebo/6.9/meshtofuel.html

Reply:
The SDF exporter code is fine as it is. But I still suggest the above in the tutorial. Thanks!

Reply:
Thank you for the comment! I improved the description of the exporter both inside the tutorial and at the beginning of the script. I hope it is better now.